### PR TITLE
Qemu q35

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

File: src/mainboard/emulation/qemu-q35/smihandler.c [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+	switch (length) {
+	case 1:
+	case 2:
+	case 4:
+	case 8:
+		break;
+	default:
+		return -1;
+	}
```
Comment (SergiiDmytruk) [inline src/mainboard/emulation/qemu-q35/smihandler.c:53]: Wouldn't this be simpler?
```c
if (length != 1 && length != 2 && length != 4 && length != 8)
    return -1;
```

**Review by SergiiDmytruk — COMMENTED** [review]
Some checks in `rom_media.c` raise questions and they at least need clarification if not modification.

File: src/mainboard/emulation/qemu-i440fx/rom_media.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	flash_ops = mem_rdev_rw_ops;
+
+	boot_dev.base = (void *)(uintptr_t)(0x100000000ULL - CONFIG_ROM_SIZE);
+	boot_dev.rdev.ops = &flash_ops;
+	boot_dev.rdev.region.size = CONFIG_ROM_SIZE;
+
+	/* Detect what type of flash we're dealing with. */
+	ptr = (volatile char *)boot_dev.base;
+
+	/* Find first byte different than any of the commands, simplified. */
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/mainboard/emulation/qemu-i440fx/rom_media.c]: https://github.com/Dasharo/coreboot/compare/5e3e382b3eab7a85726af47392b333927b26da6c..7cdf0aa873ed020156b8ae4760afeba03997f21d#diff-66d362cfd11f6d5bb5ebfe5b89a3aaf519198d541a0c0826d8549ed83659e6f9

File: src/mainboard/emulation/qemu-i440fx/rom_media.c [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		original = read8(++ptr);
+
+	write8(ptr, CLEAR_STATUS_CMD);
+	readback = read8(ptr);
+	if ((original != CLEAR_STATUS_CMD) && (readback == CLEAR_STATUS_CMD)) {
+		printk(BIOS_DEBUG, "QEMU flash behaves as RAM\n");
+		/* Restore original content. */
+		write8(ptr, original);
+	} else {
+		/* In case original value was CLEAR_STATUS_CMD, RAM is checked again. */
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/mainboard/emulation/qemu-i440fx/rom_media.c]: https://github.com/Dasharo/coreboot/compare/5e3e382b3eab7a85726af47392b333927b26da6c..7cdf0aa873ed020156b8ae4760afeba03997f21d#diff-66d362cfd11f6d5bb5ebfe5b89a3aaf519198d541a0c0826d8549ed83659e6f9

File: src/mainboard/emulation/qemu-i440fx/rom_media.c [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			printk(BIOS_DEBUG, "QEMU flash behaves as RAM\n");
+			/* Restore original content. */
+			write8(ptr, original);
+		} else if (readback == CLEARED_ARRAY_STATUS) {
+			/* Try writing original value to test whether flash is writable. */
+			write8(ptr, WRITE_BYTE_CMD);
+			write8(ptr, original);
+			write8(ptr, READ_STATUS_CMD);
+			readback = read8(ptr);
+			write8(ptr, READ_ARRAY_CMD);
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/mainboard/emulation/qemu-i440fx/rom_media.c]: https://github.com/Dasharo/coreboot/compare/5e3e382b3eab7a85726af47392b333927b26da6c..7cdf0aa873ed020156b8ae4760afeba03997f21d#diff-66d362cfd11f6d5bb5ebfe5b89a3aaf519198d541a0c0826d8549ed83659e6f9

File: src/mainboard/emulation/qemu-i440fx/rom_media.c [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+				       "QEMU flash behaves as write-protected flash\n");
+				flash_ops.writeat = NULL;
+				flash_ops.eraseat = NULL;
+			} else {
+				printk(BIOS_DEBUG, "QEMU flash behaves as writable flash\n");
+				flash_ops.writeat = qemu_writeat;
+				flash_ops.eraseat = qemu_eraseat;
+			}
+		} else {
+			printk(BIOS_ERR, "Unexpected QEMU flash behavior\n");
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/mainboard/emulation/qemu-i440fx/rom_media.c]: https://github.com/Dasharo/coreboot/compare/5e3e382b3eab7a85726af47392b333927b26da6c..7cdf0aa873ed020156b8ae4760afeba03997f21d#diff-66d362cfd11f6d5bb5ebfe5b89a3aaf519198d541a0c0826d8549ed83659e6f9

File: src/mainboard/emulation/qemu-i440fx/rom_media.c [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * - ROM - memory mapped reads, writes are ignored (FW image mounted with
+ *   '-bios');
+ * - RAM - memory mapped reads and writes (FW image mounted with e.g.
+ *   '-device loader');
+ * - flash - memory mapped reads, write and erase possible through commands.
+ *   Contrary to physical flash devices erase is not required before writing,
+ *   but it also doesn't hurt. Flash may be split into read-only and read-write
+ *   parts, like OVMF_CODE.fd and OVMF_VARS.fd. Combined size of system firmware
+ *   must not exceed 8 MiB (FW image(s) mounted with '-drive if=pflash').
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/mainboard/emulation/qemu-i440fx/rom_media.c]: https://github.com/Dasharo/coreboot/compare/5e3e382b3eab7a85726af47392b333927b26da6c..7cdf0aa873ed020156b8ae4760afeba03997f21d#diff-66d362cfd11f6d5bb5ebfe5b89a3aaf519198d541a0c0826d8549ed83659e6f9 and also in commit message for https://github.com/Dasharo/coreboot/pull/514/commits/69e1d86a8d6eed10475edbc56272fb0b446cbd44

File: build.sh
Comment (SergiiDmytruk) [inline build.sh:1]: Noticed only now: there is `usage()` function at the top which needs to be updated.

File: payloads/external/Makefile.mk [hunk 7]
```diff
@@ -168,7 +168,7 @@ IPXE_EFI := payloads/external/iPXE/ipxe/ipxe.rom
 endif
 
 $(obj)/UEFIPAYLOAD.fd: $(DOTCONFIG) $(IPXE_EFI)
-	$(MAKE) -C payloads/external/edk2 UefiPayloadPkg \
+	$(MAKE) -C payloads/external/edk2 DasharoPayloadPkg \
```
Comment (SergiiDmytruk) [inline payloads/external/Makefile.mk:171]: This commit (or an identical one) was actually done for CI in https://github.com/Dasharo/coreboot/pull/486 as I assumed we'll just switch to `DasharoPayloadPkg` everywhere in one go.  If we want to have QEMU board to be the only user of rebased EDK for now, then need some new `Kconfig` option to set build target name and override `PAYLOAD_NAME` in `edk2/Makefile`.

**Review by SergiiDmytruk — CHANGES_REQUESTED** [review]
The C code LGTM, but I missed one part of `build.sh` script and something needs to be done about EDK package name for CI to pass.

Speaking of CI, `.github/workflows/build.yml` probably needs a new section to build Q35 board.

File: payloads/external/Makefile.mk (hunk 7 above)
Comment (krystian-hebel) [reply payloads/external/Makefile.mk]: Dropped and changed PR base.

File: build.sh
Comment (krystian-hebel) [reply build.sh:1]: Added in https://github.com/Dasharo/coreboot/compare/273055e87a67891bbb91a87b23c91568bc217abd..e1bbaf8f6b731c20f175db450b304fd26b9fd99b

**Review by SergiiDmytruk — APPROVED** [review]

**Comment by SergiiDmytruk** [comment]
By the way, OVMF package enables all possible menus even if they make no sense for QEMU, just so you could explore them.

**Comment by krystian-hebel** [comment]
> By the way, OVMF package enables all possible menus even if they make no sense for QEMU, just so you could explore them.

Added full config in https://github.com/Dasharo/coreboot/pull/514/commits/4c723a0fdd5f3436403d7c68131000c03fee6368, except for OC WDT.

**Review by SergiiDmytruk — DISMISSED** [review]

**Review by SergiiDmytruk — APPROVED** [review]

**Pull request merged**